### PR TITLE
store/copr: balance region for batch cop task (#24521)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #24521 to release-5.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/24724
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/24724:release-5.0-66c8cd96b0ea
```

---

### What problem does this PR solve?

Issue Number: close pingcap/tics#1854 

Problem Summary:

For mpp queries, if a table has more than 1 replicas, the table scan task of this table does not balance the scan region between the TiFlash nodes, so it is possible that although the regions are evenly distributed on several TiFlash nodes, the 
 scanned regions on each TiFlash nodes for a specific mpp query is very uneven
### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:
In `buildBatchCopTasks`, when all the regions to be read is collected, it will call `balanceBatchCopTask` to balance region for batch cop task.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
By add some extra log in TiDB, the balance result looks like this:
```
[2021/05/10 17:48:03.318 +08:00] [INFO] [batch_coprocessor.go:293] ["before task balance store 172.16.5.59:7640 : 527 regions,  store 172.16.5.85:7640 : 359 regions,  store 172.16.5.81:7640 : 243 regions, "]
[2021/05/10 17:48:03.320 +08:00] [INFO] [batch_coprocessor.go:299] ["after task balance store 172.16.5.81:7640 : 376 regions,  store 172.16.5.59:7640 : 377 regions,  store 172.16.5.85:7640 : 376 regions, "]
[2021/05/10 17:48:03.320 +08:00] [INFO] [batch_coprocessor.go:293] ["before task balance store 172.16.5.81:7640 : 1 regions, "]
[2021/05/10 17:48:03.320 +08:00] [INFO] [batch_coprocessor.go:299] ["after task balance store 172.16.5.81:7640 : 1 regions, "]
[2021/05/10 17:48:03.321 +08:00] [INFO] [batch_coprocessor.go:293] ["before task balance store 172.16.5.85:7640 : 80 regions,  store 172.16.5.81:7640 : 227 regions,  store 172.16.5.59:7640 : 66 regions, "]
[2021/05/10 17:48:03.322 +08:00] [INFO] [batch_coprocessor.go:299] ["after task balance store 172.16.5.81:7640 : 135 regions,  store 172.16.5.59:7640 : 103 regions,  store 172.16.5.85:7640 : 135 regions, "]
```

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- balance region for batch cop task

### CODE REVIEW

**Review by hanfei1991 — APPROVED** [review]

File: store/tikv/region_cache.go [hunk 1]
```diff
@@ -664,6 +707,64 @@ func (c *RegionCache) findRegionByKey(bo *Backoffer, key []byte, isEndKey bool)
 	return r, nil
 }
 
+// OnSendFailForBatchRegions handles send request fail logic.
+func (c *RegionCache) OnSendFailForBatchRegions(bo *Backoffer, store *Store, regionInfos []RegionInfo, scheduleReload bool, err error) {
+	metrics.RegionCacheCounterWithSendFail.Add(float64(len(regionInfos)))
+	if store.storeType != TiFlash {
+		logutil.Logger(bo.GetCtx()).Info("Should not reach here, OnSendFailForBatchRegions only support TiFlash")
```
Comment (wjhuang2016) [inline store/tikv/region_cache.go:714]: The level should be `warn` or `error`

**Pull request merged**